rust/main/hyperlane-base/src/contract_sync/mod.rs

```diff
@@ -221,29 +230,39 @@ where
         &self,
         logs: Vec<(Indexed<T>, LogMeta)>,
         stored_logs_metric: &GenericCounter<AtomicU64>,
-    ) -> Vec<(Indexed<T>, LogMeta)> {
+    ) -> (Vec<(Indexed<T>, LogMeta)>, bool) {
         let deduped_logs = HashSet::<_>::from_iter(logs);
         let logs = Vec::from_iter(deduped_logs);
 
         // Store deliveries
-        let stored = match self.store.store_logs(&logs).await {
+        let stored_logs_count = match self.store.store_logs(&logs).await {
             Ok(stored) => stored,
             Err(err) => {
                 warn!(?err, "Error storing logs in db");
                 Default::default()
             }
         };
-        if stored > 0 {
+        if stored_logs_count > 0 {
             debug!(
                 domain = self.domain.as_ref(),
-                count = stored,
+                count = stored_logs_count,
                 sequences = ?logs.iter().map(|(log, _)| log.sequence).collect::<Vec<_>>(),
                 "Stored logs in db",
             );
         }
         // Report amount of deliveries stored into db
-        stored_logs_metric.inc_by(stored as u64);
-        logs
+        stored_logs_metric.inc_by(stored_logs_count as u64);
+
+        let deduped_logs_len = logs.len();
+        if (stored_logs_count as usize) < deduped_logs_len {
```

Comment:
what would happen in this scenario:
- we get 5 logs from the RPC
- we only manage to insert 2
- we skip updating the cursor because 2 < 5
- we get 5 logs from the RPC again
- we manage to insert the other 3
- we still skip updating the cursor because 3 < 5
- we keep looping?

Reply:
As it is implemented in this PR, yes.

Reply:
then this sounds like a bug? The indexer would become permanently stuck in this infinite loop whenever one message isn't inserted successfully

Reply:
I see what you mean. `store_logs` is not idempotent in sense how many logs it inserts.

Reply:
I implemented your suggestion @daniel-savu with retrying `store_logs`.